draft-dnsop-deleg.md

```diff
@@ -259,32 +260,68 @@ As defined in the SVCB document {{?RFC9460}}, the TargetName provides the hostna
+1. Use PKI for authentication, and treat any DANE-only endpoint as incompatible.  In this behavior, compatibility with `tlsa=disabled` endpoints is REQUIRED.
+1. Support both DANE and PKI for authentication, preferring DANE if it is available for each endpoint.
+
+This SvcParamKey MAY be used in any SVCB context where TLSA usage is defined.
+
+~~~
+;; The server is only authenticatable via DANE.  Any resolver that
+;; lacks DANE support will treat this record as incompatible.
+alpn=doq tlsa="..." mandatory=alpn
+
+;; The server is only authenticatable via DANE, but also offers Do53
+;; services.  All resolvers can access it, but only resolvers with
+;; DANE support will use DoQ.
+alpn=doq tlsa="..."
+
+;; The server is authenticatable via PKI.  If TLSA records exist at
+;; the SVCB-DANE owner names, it is also authenticatable via DANE.
+;; This record cannot be used to serve the zone containing its TargetName.
+alpn=doq
+
+;; The server is only authenticatable via PKI.  Any resolver that
+;; lacks PKI validation support will treat this record as incompatible.
+alpn=doq tlsa=disabled mandatory=alpn
+~~~
+{: title="tlsa SvcParam Examples for DELEG"}
 
-* "do53": indicates that a server supports plaintext, unencrypted DNS traffic over UDP or TCP as defined in {{?RFC1035}} and {{?RFC1034}} and the updates to those RFCs.
-* "dot": indicates that the server supports encrypted DNS traffic over DNS-over-TLS as defined in {{?RFC7858}}.
-* "doh": indicates that the server supports encrypted DNS traffic over DNS-over-HTTPS as defined in {{?RFC8484}}. Records that use the DoH service form may be further redirected with HTTPS resource records in the delegated zone.  The DoH path is specified with a dohpath SvcParam as specified in {{?RFC9461}}.
-* "doq": indicates that the server supports encrypted DNS traffic via DNS-over-QUIC Connections as defined in {{?RFC9250}}
+#### "ds"
 
-The order of the keys in the list dictate the order in which the nameserver SHOULD be contacted. The client SHOULD compare the order of available transports with the set of transports it supports to determine how to contact the selected nameserver.
+The "ds" SvcParamKey is a delegation parameter representing a DS RRset.  For usage notes, see {{dnssec}}.
 
-The presentation format of the SvcParamValue is a comma delimited quoted string of the available transport names. The wire format for the SvcParamValue is a string of 16-bit integers representing the TransportKey values as described in the "DELEG Transport Parameter Registry".
+The SvcParamValue is a non-empty value-list.  The presentation and wire format of each value is the same as the presentation and wire format described for the DS record as defined in {{?RFC4034}}, sections 5.3 and 5.1 respectively.  When computing the digest, the DNSKEY Owner Name is always set to "." (i.e., the root), because this DS record approves the use of the specified DNSKEY on any zone that is delegated to this endpoint.
```

Comment:
This has a risk of implementation bugs, where the DS record ends up really being used for the root (or added as a trust anchor somehow). Let's see if that can be avoided without changing much.

Labels have a length octet followed by that number of bytes. The root is `\x00`. The label length only uses the 6 lower-order bits of the length octet, with the two highest bits not used (required to be `00`). These two bits could be used to disambiguate this pseudo-"owner name" (which it really is!) from the conventional root label.

When computing the digest, this would mean setting the DNSKEY owner name to `0x20` or `0x40` (or the sum).

Reply:
My goal with this arrangement was to enable this value to be constructed using existing DS Digest codepaths that require a valid Owner Name as input (e.g. dnspython's [make_ds](https://dnspython.readthedocs.io/en/latest/dnssec.html#dns.dnssec.make_ds).  Using a byte sequence that cannot collide with a valid name is a cool idea, but it would rule out this implementation path.

I don't really see the risk that you're describing, but here are some other possible solutions:

* Use a virtual owner name of "_deleg.".
* Use a virtual owner name of the DELEG or SVCB record's owner name.
* Use a virtual owner name of the DELEG or SVCB record's TargetName.
* Omit the owner name entirely from the digest computation (deviating from the DS construction).

Reply:
I agree about the goal. Changing `make_ds` to allow this would probably be straightforward though. But anyway, perhaps best to avoid. (That also rule out your last proposal.)

Using the DELEG's owner name is a good idea, I think! (It might collide with a zone cut if present at that name.)